### ISSUE TITLE
Module aliasing duplicates instances.

### ISSUE THREAD

**Issue by Lysxia** [coq/coq#6731]
On Coq 8.6 and 8.7

I'd like to rename some module M to N. However, using "Module N := M.", instance resolution will try every typeclass instance in M a second time in N, commonly resulting in exponential-time behavior.

In the minimal example below, shouldn't both "Check" commands to produce the same debug output?

---

```coq
Module M.
  Class C (n : nat) := {}.
  Instance CS (n : nat) `{C n} : C (S n).
  Definition f (n : nat) `{C n} : unit := tt.
End M.

Set Typeclasses Debug.

(* simple apply M.CS *)
Check (M.f 1).

Module N := M.

(* simple apply N.CS
   simple apply M.CS *)
Check (N.f 1). 
```

**Comment by Blaisorblade** [comment]
To understand the desired semantics, "superglobal" effects shouldn't be duplicated by module aliasing?

**Comment by ppedrot** [comment]
No, it's a deeper problem than that, because importing two instances of morally the same module trigger the same issue. On a hopeful note, though, I'm currently working on a series of patches that would make this easy to fix.

**Comment by charguer** [comment]
Thanks for working on this.
Would it help (or only add to the confusion) to introduce a notion of "module alias"? (Hasn't OCaml followed this route?)
